README.md
# ClusterScan
ClusterScan is a tool to search for genomic clusters starting from genomic feature locations and their annotations. It allow the user to scan an annotation file (BED format with locations of specific features such as gene, transcripts, regulatory regions or anything that can be mapped on a genome) and get clusters coordinates in output. In order to build annotated clusters, ClusterScan need an additional two-columns file, storing the feature ids and the corresponding categorical information (such as Gene Ontology classes, KEGG, Pfam accessions, etc). The user can also define a custom set of categories, which makes ClusterScan very flexible.

***********************

- [How it works](#how-it-works)
- [Local installation](#local-installation)
	- [Installing Bedtools](#installing-bedtools)
	- [Installing R and the ggplot2 library](#installing-r-and-the-ggplot2-library)
	- [Installing the required Python libraries](#installing-the-required-python-libraries)
	- [Installing ClusterScan](#installing-clusterscan)
- [Docker installation](#docker-installation)
	- [Installing docker engine](#installing-docker-engine)
	- [Installing ClusterScan image](#installing-clusterscan-image)
	- [Running ClusterScan through docker](#running-clusterscan-through-docker)
- [Options](#options)
- [Input/Output](#inputoutput)
- [Other resources](#other-resources)

***********************

## How it works:
ClusterScan can use two different algorithms that perform the search and it also offer filters to select the minimum number of features to call a cluster (that can’t be a number lower than 2). The two algorithms are:

- **clusterdist**: scans the features using bedtools merge in order to find those features, for each category, which are separated by a maximum distance in base pairs that can be selected by the user (p: --distance). Some studies based on gene families, for example in human and mouse, have estimated genes within 500 Kb to be in cluster ([Niimura et al. 2003](https://www.ncbi.nlm.nih.gov/pubmed/14507991); [Tadepally et al. 2008](https://www.ncbi.nlm.nih.gov/pubmed/18559114)).

- **clustermean**: divides the genome in sliding windows and calculates, for each category, the mean number of features and the standard deviation for each accession. After that, clustermean searches for those windows showing a Z-score bigger than a given value, i.e. containing a number of features higher or equal to the relation _mean+n*stdv_ in which *n* can be set by the user (p: --seed). Thus, the seed parameter set the number of standard deviations to identify a window which serves as the beginning of the cluster. After this step, the algorithm similarly tries to extend the cluster in both the directions starting from the seed using the same relation _mean+n*stdv_ in which *n* can be set again by the user (p: --extension). Thus, the extension parameter set the number of standard deviations to identify the window(s) which serve to extend the cluster. Lastly, clustermean trims the clusters in order to replace the cluster start/end represented by the most 5' feature start and the most 3' feature end respectively.

## Local installation:
ClusterScan can be installed locally or used in a docker container. Continue reading below to locally install the tool on your own machine or you can find docker instructions [here](https://github.com/pyrevo/ClusterScan/wiki/Docker-Installation). ClusterScan requires [Python](https://www.python.org/downloads/release/python-2714/) (v2.7.x). [Bedtools](https://github.com/arq5x/bedtools2) (v2.25.0+) and [R](https://www.r-project.org/) (v3.0.0+) are needed to be in the user path. In order to draw high quality clusters distributions for features in the top 10 clusters (by number of features) found, it is also required to install the R library [ggplot2](http://ggplot2.org/) (v2.0.0+).

### Installing Bedtools
To install bedtools you need to download the package via GitHub and compile from source:
```
wget https://github.com/arq5x/bedtools2/releases/download/v2.26.0/bedtools-2.26.0.tar.gz
tar -zxvf bedtools-2.26.0.tar.gz
cd bedtools2
make
```
Once you have finished you can add bedtools to your path opening a new terminal and typing:
```
sudo nano ~/.bashrc
```
scroll until you reach the bottom of the page and add the following string:
```
export PATH=$PATH:/<path-to-bedtools>/bedtools2/bin
```
Save and exit. To test your bedtools installation you must open a new terminal and type:
```
bedtools -h
```
More informations about the bedtools installation can be found [here](http://bedtools.readthedocs.io/en/latest/content/installation.html).

### Installing R and the ggplot2 library
To install R simply type:
```
sudo apt install r-base r-base-dev
```
and enter the program with the command:
```
R
```
now you can install ggplotw with:
```
install.packages("ggplot2")
```
Finally you can test the ggplot installation with:
```
library(ggplot2)
```

### Installing the required Python libraries
ClusterScan needs few other Python libraries which can be easily installed via [pip](https://pip.pypa.io/en/stable/installing/) from [PyPI](https://pypi.python.org/pypi). As first step install python setuptools and pip on your distribution:
```
sudo apt install python-pip
sudo apt install python-setuptools
```
Then install the required dependencies via pip typing:
```
pip install pandas==0.19.1
pip install rpy2==2.8.3
pip install pybedtools==0.7.8
```
You can find more informations at: 
- [pandas](https://pandas.pydata.org/)
- [pybedtools](https://daler.github.io/pybedtools/)
- [rpy2](https://rpy2.readthedocs.io/en/version_2.8.x/)

### Installing ClusterScan
Now, you can download and extract ClusterScan in your preferred directory:
```
wget https://github.com/pyrevo/ClusterScan/archive/master.zip
unzip master.zip
```
Finally, you can add ClusterScan in your path by editing your bashrc file:
```
sudo nano ~/.bashrc
export PATH=$PATH:/<path-to-clusterscan>/ClusterScan-master
```
Save and exit. To test the clusterscan installation you must open a new terminal and type:
```
clusterscan.py -h
```
If you visualize the ClusterScan help page, you can proceed to run your first analysis by following the [ClusterScan Tutorial](https://github.com/pyrevo/ClusterScan/wiki/ClusterScan-Tutorial).

We have tested ClusterScan on [Ubuntu](https://www.ubuntu.com/) (v14.04LTS+). ClusterScan is distributed under the [GNU General Public License (GPL) Version 3](https://www.gnu.org/licenses/gpl-3.0.en.html).

## Docker installation:
ClusterScan can also be used in a docker container. Docker is an open source project intended to provide tools to performs operating-system-level virtualization also known as containerization. More information can be found [here](https://www.docker.com/what-docker). In order to use ClusterScan in a docker container, you must install the docker engine and the docker ClusterScan image. You can find a complete Linux installation guide of the docker engine following this [link](https://docs.docker.com/install/linux/docker-ce/ubuntu/).

### Installing docker engine
If you already have the docker engine installed on your machine, you can skip this step. To install the docker engine you simply have to type:
```
curl -fsSL https://get.docker.com/ | sh
```
If you haven't curl installed, you can obtain it by typing:
```
sudo apt update
sudo apt install curl
```
Once you have the docker engine installed, you may want guarantee administrator permissions to your user:
```
sudo usermod -aG docker <your-user>
```
After this step you must exit and open a new terminal. You can test your docker engine installation with the command:
```
docker run hello-world
```

### Installing ClusterScan image
To obtain the latest ClusterScan docker image type:
```
docker pull sangeslab/clusterscan:latest
```
You can test your docker ClusterScan installation by typing:
```
docker images
```
obtaining something like that:

|REPOSITORY|TAG|IMAGE ID|CREATED|SIZE|
|---|---|---|---|---|
|sangeslab/clusterscan|latest|1bc62465dc7c|3 days ago|1.44GB|
|hello-world|latest|f2a91732366c|4 months ago|1.85kB|

To get rid of the ClusterScan image:
```
docker rmi -f sangeslab/clusterscan
```
you can also remove the created containers by tracking their IDs:
```
docker ps -a
docker rm <your container ID>
```

### Running ClusterScan through docker
To run ClusterScan type:
```
docker run --user $(id -u) -it -v /full/path/to/data/folder:/data sangeslab/clusterscan bash
```
and substitute _/full/path/to/data/folder_ with the complete path to the directory containing your input data.

You can test whether CLusterScan is ready with:
```
clusterscan.py -h
```

## Options:
ClusterScan provides different options in order to finely tune the cluster search. Some of them are algorithm-specific whereas other are in common between clusterdist and clustermean:

```
Usage:
  clusterscan.py clusterdist FEATURES ANNOTATION [-o PATH] [-a NAME] [-c LIST] [--info FILE] [--singletons] [-n=<n>] [-d=<bp>]
  clusterscan.py clustermean FEATURES ANNOTATION [-o PATH] [-a NAME] [-c LIST] [--info FILE] [--singletons] [-n=<n>] [-w=<bp>] [-s=<bp>] [-k=<n>] [-e=<n>]
  clusterscan.py (-h | --help)
  clusterscan.py --version

clusterdist:
  -d, --dist=<bp>	Maximum distance between features in bp [default: 500000].

clustermean:
  -w, --window=<bp>	Window size [default: 500000].
  -s, --slide=<bp>	Sliding size [default: 250000].
  -k, --seed=<n>	Number of standard deviations to identify a window which serves as the beginning of the cluster [default: 3].
  -e, --extension=<n>	Number of standard deviations to identify the window(s) which serve to extend the cluster [default: 2].

shared options:
  -o, --output PATH	Specify output path [default: current directory].
  -a, --analysis NAME	Specify optional analysis name for output files.
  -n, --nf=<n>		Minimum number of features per cluster [default: 2].
  -c, --category LIST   Comma separated list of one or more specific categories to be analyzed [e.g. PF00001,PF00002].
                        Useful when you need to perform the analysis only for specific categories in the ANNOTATION file.
  --info FILE           Specify optional file to describe categories.
  --singletons          Identify singletons after clusters and bystanders annotation.
```

An example of execution with clusterdist:
```
clusterscan.py clusterdist my_genes.bed my_categories.txt -d 250000 -a analysis_01
```

An example of execution with clustermean:
```
clusterscan.py clustermean my_genes.bed my_categories.txt --info my_descriptions.txt -w 10000 -s 5000
```

## Input/Output:
ClusterScan is composed by three files: _clusterscan.py_, _algos.py_ and _docopt.py_. The first represent the tool itself, the second stores the functions belonging to the two algorithms described above, and the latter is a dependence that provides an interface for the command-line menu. The three files need to be in the same folder.

ClusterScan requires two mandatory files and a third optional file in input. The first mandatory file is represented by a six-field bed file (FEATURES) containing the location of the features for which to build cluster. It must follow the organization of the [BED file format](https://genome.ucsc.edu/FAQ/FAQformat.html#format1). The second file is a two-column table (ANNOTATION) containing the name of the features as first column (must correspond to the name of the feature in the FEATURES file) and a string representing the categorical information used in the cluster search as second column. This "category" can be represented for esample by Gene Ontology classes, KEGG, Pfam accessions etc. The user can also search for clusters based on a custom made series of categories describing them using an optional additional two-column tab-delimited .txt file (category, description) that can be read by the program (p: --info). Moreover, the user can pass to the program a comma separated list of categories on which perform the analysis (p: --category) in the case in which he/she does not want to analyze all the categories in the ANNOTATION file. It is not uncommon that a feature is associated with different accessions at the same time and, in this case, the feature can be part of multiple different clusters. The user can also choose to identify features which exist as _singletons_ (p: --singletons). Those represent features belonging to a certain category, for which at least one cluster was annotated, that are outside of any of the corresponding clusters.

ClusterScan gives six or seven files in output depending on whether the user has chosen or not to identify the singletons:

|File|Description|
|---|---|
|	clusters.tsv | stores the coordinates of all the clusters found. It contains an unique ID for each cluster; the category to which the cluster belongs; the chromosome/scaffold (chr) on which it resides; its start and end coordinates; the number of features within the cluster (n_features) and the number of features which overlap the cluster but belong to a different category (n_bystanders). |
|		clusters.bed | a bed version of the previous file for compatibility with other tools for downstream analysis. Fields respect the bed format and are (from first to last column): chromosome/scaffold on which the cluster resides; its start/end coordinates using a 0-based start and 1-based end system of coordinates; the cluster ID; the number of features within the cluster; the strand; the category describing the cluster. |
|		summary.tsv | contains a category based summary of the cluster analysis with the category; the number of clusters found for each category (n_clusters); the total number of features found for clusters belonging to each category (n_ft); the number of bystanders (n_bs); the number of features and bystanders found in the clusters which contains the minimal and the maximal number of these features (max_ft and min_ft / max_bs and min_bs) for each category. |
|		features.tsv | a list of features found to be in overlap with clusters. It contains the chromosome/scaffold (chr) on which the feature resides; its start and end coordinates; the feature name; the strand; the ID of the cluster to which it overlaps; the category to which the cluster (and the feature) belongs. |
|		bystanders.tsv | a list of bystanders found to be in overlap with clusters. It contains the chromosome/scaffold (chr) on which the feature resides; its start and end coordinates; the feature name; the strand; the ID of the cluster to which it overlaps; the category to which the cluster (but no the feature) belongs. |
|		distribution.pdf | an histogram which shows the distribution of features in the top-10 categories by number of features. |
|		singletons.tsv | a list of singletons found to be out from the annotated clusters for each category. It contains the chromosome/scaffold (chr) on which it resides; its start and end coordinates; the feature name; the strand; the category to which the feature belongs. |

The program assigns an unique ID for each cluster during the analysis, joining the “C” letter with a progressive number starting from 1. The user can trace back a specific cluster at any time in the different output files simply looking at the ID reported in the corresponding column.

## Other resources:
You can contact us by opening a new issue on GitHub or asking a new question on our dedicated [mailing list](https://groups.google.com/forum/#!forum/clusterscan-support). Take a look at our Wiki [here](https://github.com/pyrevo/ClusterScan/wiki/Welcome-to-the-ClusterScan-wiki!).


CHANGES.md
# Changelog
All notable changes to this project will be documented in this file.

## 0.3.0 - 2021-08-25
### Added
- .gitignore file.
- environment.yml for conda env configuration.

## 0.2.2 - 2018-03-20
### Changed
- NA instead of 0 when one category is used for the search.

## 0.2.1 - 2018-01-03
### Added
- possibility to pass a list of categories to be analyzed (p: --category).
- identification of singletones (e.g. those features outside any cluster of the category to which the feature belongs; p: --singletons).

### Changed
- ClusterScan nomenclature regarding the categorical annotation.

## 0.2.0 - 2017-12-22
### Added
- analysis duration benchmark in standard output.
- summmary of the analysis in standard output.
- "CHANGES.md" and "README.md".
environmentenvironment
### Changed
- improve _clustermean_ speed.

### Removed
- "CHANGES.txt" and "README.txt".

## 0.1.0 - 2017-05-16
### Added
- first ClusterScan release.


clusterscan.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright (C) 2017-2018 Massimiliano Volpe and Marco Miralto

# This file is part of ClusterScan.

# ClusterScan is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# ClusterScan is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with ClusterScan.  If not, see <http://www.gnu.org/licenses/>.


"""Description:
  ClusterScan, search for clusters of features in a given annotation.

Usage:
  clusterscan.py clusterdist FEATURES ANNOTATION [-o PATH] [-a NAME] [-c LIST] [--info FILE] [--singletons] [-n=<n>] [-d=<bp>]
  clusterscan.py clustermean FEATURES ANNOTATION [-o PATH] [-a NAME] [-c LIST] [--info FILE] [--singletons] [-n=<n>] [-w=<bp>] [-s=<bp>] [-k=<n>] [-e=<n>]
  clusterscan.py (-h | --help)
  clusterscan.py --version

Options:
  -h, --help                        Show this screen.
  -o, --output PATH                 Specify output path [default: ./].
  -a, --analysis NAME               Specify optional analysis name for output files.
  -n, --nf=<n>                      Minimum number of features per cluster [default: 2].
  -d, --dist=<bp>                   Maximum distance between features in bp [default: 500000].
  -w, --window=<bp>                 Window size [default: 500000].
  -s, --slide=<bp>                  Sliding size [default: 250000].
  -k, --seed=<n>                    Number of standard deviations to identify a window which serves as the beginning of the cluster [default: 3].
  -e, --extension=<n>               Number of standard deviations to identify the window(s) which serve to extend the cluster [default: 2].
  -c, --category LIST               Comma separated list of one or more specific categories to be analyzed [e.g. PF00001,PF00002].
                                    Useful when you need to perform the analysis only for specific categories in the ANNOTATION file.
  --info FILE                       Specify optional file to describe categories.
  --singletons                      Identify singletons after clusters and bystanders annotation.
  --version                         Show program version.
"""

import time
import os
import warnings

import pandas as pd
import pybedtools
from docopt import docopt
from rpy2 import robjects
from rpy2.robjects import pandas2ri
from rpy2.robjects.packages import importr

from algos import *

with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    import rpy2.robjects.lib.ggplot2 as ggplot2

start_time = time.time()


def input_tester(file_path):
    """Check for the presence of input files."""
    try:
        open(file_path)
    except IOError:
        raise SystemExit('Unable to open %s, file does not exist!' % (file_path.split('/')[-1]))
    else:
        pass


def options_tester(option, n, string):
    """Check if rules for parameters are respected."""
    if option < n:
        raise ValueError(string)
    else:
        pass


def rpy2_plotter(anno, clusters, name):
    """Plot genes distribution in clusters using ggplot2 from R."""
    pandas2ri.activate()
    grdevices = importr('grDevices')
    rprint = robjects.globalenv.get("print")

    anno = anno.sort_values(by="n_ft", ascending=False)
    anno = anno.head(n=10)
    category = anno["category"].tolist()
    clusters = clusters[clusters["category"].isin(category)]
    clusters = pandas2ri.py2ri(clusters)

    pp = ggplot2.ggplot(clusters) + ggplot2.aes_string(x="n_features") + ggplot2.geom_histogram(binwidth=1) + ggplot2.facet_wrap(robjects.Formula("~category"), ncol=5) + ggplot2.labs(x="Number of Features", y="Number of Clusters", title="Clusters distribution")

    grdevices.pdf(file=name, width=11.692, height=8.267)
    rprint(pp)
    grdevices.dev_off()


def main():
    # test for input files availability
    input_tester(arguments['FEATURES'])
    input_tester(arguments['ANNOTATION'])

    if arguments['--info'] is None:
        pass
    else:
        input_tester(arguments['--info'])

    # clusters can't contain less than 2 features
    error1 = "Minimum number of features per cluster must be a number higher than 1!"
    options_tester(int(arguments['--nf']), 2, error1)

    # window size can't be lower than sliding size
    error2 = "Sliding size can't be higher than window size!"
    options_tester(int(arguments['--window']), int(arguments['--slide']), error2)

    # window size can't be lower than sliding size
    error3 = "Seed or extension can't be a number lower than 1!"
    options_tester(int(arguments['--seed']), 1, error3)
    options_tester(int(arguments['--extension']), 1, error3)

    # build database
    feat = pd.read_table(arguments['FEATURES'], header=None, usecols=range(6), dtype={0: str})
    anno = pd.read_table(arguments['ANNOTATION'], header=None)

    feat.columns = ['chr', 'start', 'end', 'name', 'score', 'strand']
    anno.columns = ['name', "category"]
    # anno["category"] = anno["category"].fillna("Unknown")
    n = list(feat.name.unique())

    # pdtable stores genes annotation and corresponding categories
    pdtable = pd.merge(feat, anno, on='name', how='outer')
    pdtable["category"] = pdtable["category"].fillna("Unknown")
    pdtable = pdtable[pd.notnull(pdtable["category"])]
    pdtable = pdtable[pd.notnull(pdtable['chr'])]
    pdtable[['start', 'end']] = pdtable[['start', 'end']].astype(int)
    pdtable = pdtable.drop_duplicates(['name', "category"])
    # movq print str(pdtable)
    all_features = pdtable
    pdtable = pdtable[pdtable["category"] != "Unknown"]

    # list unique categories
    if arguments['--category'] is None:
        l = list(pdtable.category.unique())
    else:
        l = arguments['--category'].split(',')
    # test the argument
    if set(l) <= set(list(pdtable.category.unique())):
        pass
    else:
        raise SystemExit('Some categories passed through the -c parameter are not present in the input files. Please, check your list and run the analysis again.')

    # inizialize empty table to be filled with clusters
    table = pd.DataFrame()

    # choose the algorithm
    if arguments['clusterdist'] is True:
        print "ClusterScan is running with clusterdist..."

        # movq: arguments should be changed
        table = do_clusterdist(l,pdtable, table, arguments)
    else:
        print "ClusterScan is running with clustermean..."

        table = do_clustermean(l, pdtable, table, arguments)

    if table.empty:
        print "ClusterScan didn't found any cluster!"
        exit()
    else:
        pass

    # generate cluster table and filter it
    table.columns = ["chr", "start", "end", "n_features", "category"]
    table = table.sort_values(["category", "chr"], ascending=[True, True])
    table = table[table["n_features"] >= int(arguments['--nf'])]
    table = table.sort_values(by=["category"], ascending=[True])
    # table["cluster_id"] = range(1, len(table) + 1)
    table["cluster_id"] = ["C"+str(i) for i in range(1, len(table) + 1)]
    # get the total number of clusters
    c = table.shape[0]

    # generate output of clusters in BED format
    bedTbl = table.copy()
    bedTbl["strand"] = "+"
    bedTbl = bedTbl[[0, 1, 2, 5, 3, 6, 4]]
    bed = pybedtools.BedTool().from_dataframe(bedTbl).sort()

    # generate table of features by intersect feature with clusters
    all_features_bed = pybedtools.BedTool().from_dataframe(all_features)
    clusters = pybedtools.BedTool().from_dataframe(table)
    features = all_features_bed.intersect(clusters, wb=True)
    features = pd.read_table(features.fn, header=None, dtype={0: str, 7: str})
    cl_features = features[features[6] == features[11]]
    cl_features = cl_features[[0, 1, 2, 3, 4, 5, 12, 11]]
    cl_features.columns = ["chr", "start", "end", "name", "score",
                           "strand", "cluster_id", "category"]
    cl_features.drop('score', axis=1, inplace=True)
    # generate table of bystanders
    bystanders = features[features[6] != features[11]]

    # comment if you want to search for bystanders using only 1 category
    #if len(l) == 1:
    #    bystanders = pd.DataFrame()
    #else:
    #    pass

    # control for bystander = 0 (when program run with 1 category)
    if bystanders.empty:
        table = table[[5, 4, 0, 1, 2, 3]]
        table["n_bystanders"] = 0
    else:
        bystanders = bystanders[[0, 1, 2, 3, 4, 5, 12, 11]]
        bystanders.columns = ["chr", "start", "end", "name", "score",
                              "strand", "cluster_id", "category"]
        # prevent bystanders with 2+ different categories to be counted twice
        bystanders = bystanders.drop_duplicates(['name', "cluster_id"])
        # prevent features with 2+ different categories to be bystanders in theyr clusters
        bs_merge = pd.merge(bystanders, cl_features, how='outer', indicator=True)
        bystanders = bs_merge.ix[bs_merge._merge == 'left_only']
        #bystanders = bystanders.drop(bystanders.columns[8], axis=1)
        bystanders = bystanders.drop(bystanders.columns[[4, 8]], axis=1)
        # count bystanders number
        bs_count = bystanders.groupby("cluster_id").count().reset_index()
        bs_count = bs_count[[0, 1]]
        bs_count.columns = ["cluster_id", "n_bystanders"]
        table = table.merge(bs_count, on="cluster_id", how='outer')
        table = table[[5, 4, 0, 1, 2, 3, 6]]
        table["n_bystanders"].fillna(0, inplace=True)
        table["n_bystanders"] = table["n_bystanders"].astype(int)
        bystanders["start"] += 1


    # generate summary table
    summary = table.drop(table.columns[[0, 2, 3, 4]], axis=1)
    # calculate total number of clusters per-category
    n_clusters = summary.groupby("category").count().reset_index()
    n_clusters = n_clusters.drop("n_bystanders", axis=1)
    n_clusters.columns = ["category", "n_clusters"]
    # calculate total number of features and bystanders per-category
    n_ft_bs = summary.groupby("category").sum().reset_index()
    n_ft_bs.columns = ["category", "n_ft", "n_bs"]
    # calculate maximum number of features and bystander in cluster
    max_ft_bs = summary.groupby("category").max().reset_index()
    max_ft_bs.columns = ["category", "max_ft", "max_bs"]
    # calculate minimum number of features and bystander in cluster
    min_ft_bs = summary.groupby("category").min().reset_index()
    min_ft_bs.columns = ["category", "min_ft", "min_bs"]
    # add category description if an info file is provided
    if arguments['--info'] is None:
        summary = n_clusters.merge(n_ft_bs, on="category").merge(max_ft_bs, on="category").merge(min_ft_bs, on="category")
    else:
        desc = pd.read_table(arguments['--info'], header=None)
        desc.columns = ["category", "description"]
        summary = n_clusters.merge(n_ft_bs, on="category").merge(max_ft_bs, on="category").merge(min_ft_bs, on="category").merge(desc, on="category")

    # assign file names and save tables as result
    if not os.path.exists(arguments['--output']):
        os.makedirs(arguments['--output'])

    if arguments['--analysis'] is None:
        feat_name = os.path.join(arguments['--output'], 'features.tsv')
        byst_name = os.path.join(arguments['--output'], 'bystanders.tsv')
        clus_name = os.path.join(arguments['--output'], 'clusters.tsv')
        summ_name = os.path.join(arguments['--output'], 'summary.tsv')
        bed_name = os.path.join(arguments['--output'], 'clusters.bed')
        plot_name = os.path.join(arguments['--output'], 'distribution.pdf')
    else:
        feat_name = os.path.join(arguments['--output'], arguments['--analysis']+'_features.tsv')
        byst_name = os.path.join(arguments['--output'], arguments['--analysis']+'_bystanders.tsv')
        clus_name = os.path.join(arguments['--output'], arguments['--analysis']+'_clusters.tsv')
        summ_name = os.path.join(arguments['--output'], arguments['--analysis']+'_summary.tsv')
        bed_name = os.path.join(arguments['--output'], arguments['--analysis']+'_clusters.bed')
        plot_name = os.path.join(arguments['--output'], arguments['--analysis']+'_distribution.pdf')

    cl_features["start"] += 1
    table["start"] += 1

    if bystanders.empty:
        table["n_bystanders"] = 'NA'
        summary["n_bs"] = 'NA'
        summary["max_bs"] = 'NA'
        summary["min_bs"] = 'NA'

    cl_features.to_csv(feat_name, sep='\t', header=True, index=False)
    bystanders.to_csv(byst_name, sep='\t', header=True, index=False)
    table.to_csv(clus_name, sep='\t', header=True, index=False)
    summary.to_csv(summ_name, sep='\t', header=True, index=False)

    if arguments['--analysis'] is None:
        bed.saveas(bed_name, trackline='track name="%s" description="chr start end cluster_id n_features strand category"' % (arguments['FEATURES']))
    else:
        bed.saveas(bed_name, trackline='track name="%s" description="chr start end cluster_id n_features strand category"' % (arguments['--analysis']))

    # plot a duistribution for top 10 clusters (per n of features)
    rpy2_plotter(summary, table, plot_name)

    if arguments['--singletons'] is True:
        print "Singletons identification has been launched..."
        singletons = pd.DataFrame()
        singletons = do_singletons(l, pdtable, bedTbl, singletons, arguments)
        if singletons.empty:
            print "ClusterScan didn't found any singleton!"
        else:
            if arguments['--analysis'] is None:
                st_name = os.path.join(arguments['--output'], 'singletons.tsv')
            else:
                st_name = os.path.join(arguments['--output'], arguments['--analysis']+'_singletons.tsv')

            singletons.columns = ["chr", "start", "end", "name", "score",
                              "strand", "category"]
            singletons.drop('score', axis=1, inplace=True)
            singletons["start"] += 1
            singletons.to_csv(st_name, sep='\t', header=True, index=False)
    else:
        pass

    print '\n%s\t%s' % ("Total number of unique features scanned:", len(n))
    print '%s\t%s' % ("Total number of unique categories scanned:", len(l))
    print '%s\t%s\n' % ("Total number of clusters found:", c)


# program execution
if __name__ == '__main__':
    arguments = docopt(__doc__, version='ClusterScan 0.3.0')
    # print arguments
    main()
    print "--- %s seconds ---" % (int(round(time.time() - start_time, 0)))
